kyo-stm/shared/src/main/scala/kyo/STM.scala

```diff
@@ -0,0 +1,180 @@
+      *
+      * @param isolate
+      *   The isolation scope for the transaction
+      * @param retrySchedule
+      *   The schedule for retrying failed transactions
+      * @param v
+      *   The transactional computation to run
+      * @return
+      *   The result of the computation if successful
+      */
+    def run[E, A: Flat, S](isolate: Isolate[S], retrySchedule: Schedule = defaultRetrySchedule)(v: A < (STM & Abort[E] & Async & S))(
+        using frame: Frame
+    ): A < (S & Async & Abort[E | FailedTransaction]) =
+        isolate.use { st =>
+            run(retrySchedule)(isolate.resume(st, v)).map(isolate.restore(_, _))
+        }
+
+    /** Executes a transactional computation with default retry behavior. This version only supports Abort and Async effects within the
+      * transaction, but provides a simpler interface when additional effect isolation is not needed.
+      *
+      * @param v
+      *   The transactional computation to run
+      * @return
+      *   The result of the computation if successful
+      */
+    def run[E, A: Flat](v: A < (STM & Abort[E] & Async))(using frame: Frame): A < (Async & Abort[E | FailedTransaction]) =
+        run(defaultRetrySchedule)(v)
+
+    /** Executes a transactional computation with custom retry behavior. Like the version above, this only supports Abort and Async effects
+      * but allows configuring how transaction conflicts are retried.
+      *
+      * @param retrySchedule
+      *   The schedule for retrying failed transactions
+      * @param v
+      *   The transactional computation to run
+      * @return
+      *   The result of the computation if successful
+      */
+    def run[E, A: Flat](retrySchedule: Schedule)(v: A < (STM & Abort[E] & Async))(
+        using frame: Frame
+    ): A < (Async & Abort[E | FailedTransaction]) =
+        useTid {
+            case -1L =>
+                // New transaction without a parent, use regular commit flow
+                Retry[FailedTransaction](retrySchedule) {
+                    withNewTid { tid =>
+                        Var.runWith(RefLog.empty)(v) { (log, result) =>
+                            IO.Unsafe {
+                                // Attempt to acquire locks and commit the transaction
+                                val (locked, unlocked) =
+                                    // Sort references by identity to prevent deadlocks
+                                    log.toSeq.sortBy((ref, _) => ref.hashCode)
+                                        .span((ref, entry) => ref.lock(entry))
+
+                                if unlocked.nonEmpty then
+                                    // Failed to acquire some locks - rollback and retry
+                                    locked.foreach((ref, entry) => ref.unlock(entry))
+                                    Abort.fail(FailedTransaction(frame))
```

Comment:
`retry`?

Reply:
It's not possible to use it directly because `retry` adds a pending `STM` instead of just `Abort[FailedTransaction]`. It could return `Abort` but returning `STM` simplifies the user-facing effect tracking.